posthog/hogql_queries/insights/paths_query_runner.py

```diff
@@ -0,0 +1,468 @@
+from datetime import datetime, timedelta
+from math import ceil
+from re import escape
+from typing import Any, Dict, Literal, cast
+from typing import Optional
+
+from posthog.caching.insights_api import BASE_MINIMUM_INSIGHT_REFRESH_INTERVAL, REDUCED_MINIMUM_INSIGHT_REFRESH_INTERVAL
+from posthog.caching.utils import is_stale
+from posthog.constants import PAGEVIEW_EVENT, SCREEN_EVENT, HOGQL
+from posthog.hogql import ast
+from posthog.hogql.constants import LimitContext
+from posthog.hogql.parser import parse_select, parse_expr
+from posthog.hogql.printer import to_printed_hogql
+from posthog.hogql.property import property_to_expr
+from posthog.hogql.query import execute_hogql_query
+from posthog.hogql.timings import HogQLTimings
+from posthog.hogql_queries.query_runner import QueryRunner
+from posthog.hogql_queries.utils.query_date_range import QueryDateRange
+from posthog.models import Team
+from posthog.models.filters.mixins.utils import cached_property
+from posthog.queries.util import correct_result_for_sampling
+from posthog.schema import (
+    HogQLQueryModifiers,
+    PathsQueryResponse,
+    PathCleaningFilter,
+    PathsFilter,
+    PathType,
+)
+from posthog.schema import PathsQuery
+
+EVENT_IN_SESSION_LIMIT_DEFAULT = 5
+SESSION_TIME_THRESHOLD_DEFAULT_MILLISECONDS = 1800000  # milliseconds to 30 minutes
```

Comment:
I'd write it like this for even more clarity:

```suggestion
SESSION_TIME_THRESHOLD_DEFAULT_MILLISECONDS = 30 * 60 * 1000  # 30 minutes
```